exporters/sender/jdk/src/main/java/io/opentelemetry/exporter/sender/jdk/internal/JdkHttpSender.java

```diff
@@ -129,46 +162,64 @@ private HttpResponse<byte[]> sendInternal(Consumer<OutputStream> marshaler) {
-          || !retryableStatusCodes.contains(httpResponse.statusCode())) {
-        return httpResponse;
+      if (attempt > 0) {
+        // Compute and sleep for backoff
+        long upperBoundNanos = Math.min(nextBackoffNanos, retryPolicy.getMaxBackoff().toNanos());
+        long backoffNanos = ThreadLocalRandom.current().nextLong(upperBoundNanos);
+        nextBackoffNanos = (long) (nextBackoffNanos * retryPolicy.getBackoffMultiplier());
+        try {
+          TimeUnit.NANOSECONDS.sleep(backoffNanos);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          break; // Break out and return response or throw
+        }
+        // If after sleeping we've exceeded timeoutNanos, break out and return response or throw
+        if ((System.nanoTime() - startTimeNanos) >= timeoutNanos) {
+          break;
+        }
       }
 
-      // Compute and sleep for backoff
-      long upperBoundNanos = Math.min(nextBackoffNanos, retryPolicy.getMaxBackoff().toNanos());
-      long backoffNanos = ThreadLocalRandom.current().nextLong(upperBoundNanos);
-      nextBackoffNanos = (long) (nextBackoffNanos * retryPolicy.getBackoffMultiplier());
+      attempt++;
+      requestBuilder.timeout(Duration.ofNanos(timeoutNanos - (System.nanoTime() - startTimeNanos)));
       try {
-        TimeUnit.NANOSECONDS.sleep(backoffNanos);
-      } catch (InterruptedException e) {
-        Thread.currentThread().interrupt();
-        throw new IllegalStateException(e);
+        httpResponse = sendRequest(requestBuilder, byteBufferPool);
+      } catch (IOException e) {
+        exception = e;
       }
-      if ((System.nanoTime() - startTimeNanos) >= timeoutNanos) {
+
+      if (httpResponse != null && !retryableStatusCodes.contains(httpResponse.statusCode())) {
         return httpResponse;
       }
-    } while (true);
+      if (exception != null && !isRetryableException(exception)) {
+        throw exception;
+      }
+    } while (attempt < retryPolicy.getMaxAttempts());
+
+    if (httpResponse != null) {
+      return httpResponse;
+    }
+    throw exception;
   }
 
   private HttpResponse<byte[]> sendRequest(
-      HttpRequest.Builder requestBuilder, ByteBufferPool byteBufferPool) {
+      HttpRequest.Builder requestBuilder, ByteBufferPool byteBufferPool) throws IOException {
     try {
       return client.send(requestBuilder.build(), HttpResponse.BodyHandlers.ofByteArray());
-    } catch (IOException | InterruptedException e) {
-      if (e instanceof InterruptedException) {
-        Thread.currentThread().interrupt();
-      }
-      // TODO: is throwable retryable?
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
       throw new IllegalStateException(e);
     } finally {
       byteBufferPool.resetPool();
     }
   }
 
+  private static boolean isRetryableException(IOException throwable) {
+    return throwable instanceof HttpConnectTimeoutException;
```

Comment:
Have we thought about expanding the scope a bit to include the parent type `HttpTimeoutException`? This could allow for retries for send timeouts beyond just the connection timeout.

Reply:
So for OkHttp, there are settings for connectTimeout, readTimeout, and writeTimeout. When these settings are exceeded, you get a corresponding timeout.

For JDK HttpClient, there's only a connectTimeout setting. So I'm not actually sure what the read / write timeout settings are or how to reproduce one of these things in tests to ensure we're doing the right thing.

With that said, I do think we want to retry in all timeout scenarios, so I'm going to expand the criteria of retryable exceptions to include any `HttpTimeoutException`.